source/vstd/std_specs/hash.rs

```diff
@@ -428,6 +428,234 @@ pub fn ex_hash_map_clear<Key, Value, S>(m: &mut HashMap<Key, Value, S>)
+    Q: Hash + Eq + ?Sized,
+    S: BuildHasher,
+
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> result
+            == sets_contains_borrowed_key(m@, k),
+{
+    m.contains(k)
+}
+
+// The specification for `get` has a parameter `key: &Q` where you'd
+// expect to find `key: &Key`. This allows for the case that `Key` can
+// be borrowed as something other than `&Key`. For instance,
+// `Box<u32>` can be borrowed as `&u32` and `String` can be borrowed
+// as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a map to map a borrowed key of type
+// `&Q` to a certain value. And the postcondition of `get` says that
+// its result matches the output of that specification function. (It
+// also says that its result corresponds to the output of
+// `contains_borrowed_key`, discussed above.) But this isn't very
+// helpful by itself, since there's no body to that specification
+// function. So we have special-case axioms that say what this means
+// in two important circumstances: (1) `Key = Q` and (2) `Key =
+// Box<Q>`.
+pub spec fn sets_borrowed_key_to_key<Key, Q: ?Sized>(m: Set<Key>, k: &Q, v: &Key) -> bool;
+
+pub broadcast proof fn axiom_sets_deref_key_to_value<Q>(m: Set<Q>, k: &Q, v: &Q)
+    ensures
+        #[trigger] sets_borrowed_key_to_key::<Q, Q>(m, k, v) <==> m.contains(*k) && k == v,
+{
+    admit();
+}
+
+pub broadcast proof fn axiom_sets_box_key_to_value<Q>(m: Set<Box<Q>>, q: &Q, v: &Box<Q>)
+    ensures
+        #[trigger] sets_borrowed_key_to_key::<Box<Q>, Q>(m, q, v) <==> (m.contains(*v) && Box::new(
+            *q,
+        ) == v),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_get<'a, Key, S, Q>(m: &'a HashSet<Key, S>, k: &Q) -> (result: Option<
+    &'a Key,
+>) where Key: Borrow<Q> + Hash + Eq, Q: Hash + Eq + ?Sized, S: BuildHasher
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> match result {
+            Some(v) => sets_borrowed_key_to_key(m@, k, v),
+            None => !sets_contains_borrowed_key(m@, k),
+        },
+{
+    m.get(k)
+}
+
+// The specification for `remove` has a parameter `key: &Q` where
+// you'd expect to find `key: &Key`. This allows for the case that
+// `Key` can be borrowed as something other than `&Key`. For instance,
+// `Box<u32>` can be borrowed as `&u32` and `String` can be borrowed
+// as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively. To deal with this, we have a specification function
+// that opaquely specifies what it means for two maps to be related by
+// a remove of a certain `&Q`. And the postcondition of `remove` says
+// that `old(self)@` and `self@` satisfy that relationship. (It also
+// says that its result corresponds to the output of
+// `contains_borrowed_key` and `maps_borrowed_key_to_value`, discussed
```

Comment:
Should refer to the set equivalents instead of `contains_borrowed_key` and `maps_borrowed_key_to_value`.